Add advanced simulation tutorial (#591)
SciML/Catalyst.jl: docs/src/tutorials/advanced_simulations.md

# [Advanced Simulation Options](@id advanced_simulations)

> throughout this tutorial, we have described how to perform basic ODE, SDE, and Jump simulations of Catalyst models. However, there exists a large number of additional options for performing these simulations. For a comprehensive overview, please read the [documentation of the DifferentialEquations package](https://docs.sciml.ai/DiffEqDocs/stable/), on which Catalyst depends for all simulations. Here, we will not provide a comprehensive overview, but rather demonstrate a couple of useful features of DifferentialEquations. Finally, while DifferentialEquations have a large number of options for tuning simulation performance, these features will not be considered in this tutorial.

Reviewer: ```suggestion
Throughout the preceding tutorials, we have shown the basics of how to solve ODE, SDE, and jump process models generated from Catalyst `ReactionSystem`s. In this tutorial we'll illustrate some more advanced functionality that can be useful in many modeling contexts, and that provide conveniences for common workflows. For a comprehensive overview of solver properties, parameters, and manipulating solution objects, please read the [documentation of the DifferentialEquations package](https://docs.sciml.ai/DiffEqDocs/stable/), which Catalyst uses for all simulations. 
```

### Monte Carlo simulations using `EnsembleProblem`s

> Especially when performing stochastic simulations, one often wishes to run a large number of simulations. While it is possible to do so e.g. through a `for` loop, a more convenient way is to create a so-called `EnsembleProblem`. An advantage of these is that they contain a simple interface for modifying a problem across the ensemble, as well as automatic parallelization across CPUs. For a more throughout description, please read [the Parallel Ensemble Simulations section of the DifferentialEquation documentation](https://docs.sciml.ai/DiffEqDocs/stable/features/ensemble/#ensemble). Here, we will give a brief introduction to this feature.

Reviewer: ```suggestion
In many contexts one needs to run multiple simulations of a model, for example to collect statistics of SDE or jump process solutions, or to systematically vary parameter values within a model. While it is always possible to manually run such ensembles of simulations via a `for` loop, DifferentialEquations.jl provides the `EnsembleProblem` as a convenience to manage structured collections of simulations. `EnsembleProblem`s provide a simple interface for modifying a problem between individual simulations, and offers several options for batching and/or parallelizing simulation runs. For a more thorough description, please read [the Parallel Ensemble Simulations section of the DifferentialEquations documentation](https://docs.sciml.ai/DiffEqDocs/stable/features/ensemble/#ensemble). Here, we will give a brief introduction to the use of `EnsembleProblem`s from Catalyst-generated models.
```

> We are using a simple single-component self-activation model (which contains bistability):

Reviewer: ```suggestion
Let's look at a single-component bistable self-activation model:
```

```@example ex1
rn = @reaction_network begin
    v0 + hill(X,v,K,n), ∅ --> X
    deg, X --> ∅
end
u0 = [:X => 0.0]
tspan = (0.0,1000.0)
p = [:v0 => 0.1, :v => 2.5, :K => 75.0, :n => 2.0, :deg => 0.01];
sprob = SDEProblem(rn,u0,tspan,p)
nothing # hide
```
we can then use our `SDEProblem` as input to an `EnsembleProblem`:
```@example ex1
eprob = EnsembleProblem(sprob)
```

> next, the `EnsembleProblem` can be used as an input to the `solve()` command. It has the same options as when simulating the `SDEProblem` directly, however, it has an additional argument `trajectories` to determine how many simulations should be performed. 

Reviewer: ```suggestion
The `EnsembleProblem` can now be used as input to the `solve` command. It has the same options as when simulating the `SDEProblem` directly, however, it has an additional argument `trajectories` to determine how many simulations should be performed. 
```
To make clear it has the same input maybe select a specific solver here too?

```@example ex1
esol = solve(eprob; trajectories=5)
```

> This simulation is automatically multithreaded over all available threads. Please read [this documentation](https://docs.sciml.ai/DiffEqDocs/stable/features/ensemble/#EnsembleAlgorithms) for more information on parallelisation alternatives. The ensemble simulations can be plotted using the `plot()` function, which will by default display all trajectories:

Reviewer: ```suggestion
This simulation is automatically multithreaded over all available threads. Please read [this documentation](https://docs.sciml.ai/DiffEqDocs/stable/features/ensemble/#EnsembleAlgorithms) for more information on parallelisation alternatives. The ensemble simulations can be plotted using the `plot` function, which by default displays all trajectories:
```

```@example ex1
plot(esol)
```

> Sometimes when performing a large number of ensemble simulations, the plots get very dense. In these cases, the plotting argument is `linealpha` (which sets trajectory transparency) may be useful:

Reviewer: ```suggestion
Sometimes when performing a large number of ensemble simulations, the plots get very dense. In these cases, the plot argument `linealpha` (which sets trajectory transparency) may be useful:
```
Did you forget this argument below?

```@example ex1
esol = solve(eprob; trajectories=100)
plot(esol)
```

> Sometimes, one wishes to perform the same simulation a large number of times, while making minor modifications to the problem each time. This can be done by giving a problem function argument to the `EnsembleProblem`. Let us consider ODE simulations of a simple birth/death process:

Reviewer: ```suggestion
Sometimes, one wishes to perform the same simulation a large number of times, while making minor modifications to the problem each time. This can be done by giving a problem function, `prob_func`, argument to the `EnsembleProblem`. Let us consider ODE simulations of a simple birth/death process:
```

```@example ex1
rn = @reaction_network begin
    (b,1.0), ∅ <--> X
end
u0 = [:X => 1.0]
tspan = (0.0,1.0)
p = [:b => 1.];
oprob = ODEProblem(rn,u0,tspan,p)
nothing # hide
```
We wish to simulate this model for a large number of values of `b`. We do this by creating a `prob_func` that will make a modification to the problem at the start of each Monte Carlo simulation:
```@example ex1
b_values = 1.0:0.1:2.0
function prob_func(prob,i,repeat)

>     remake(prob; p=[b_values[i]])

Reviewer: This should use a parameter mapping and not rely on an implicit ordering (even if it is just one state, we don't want to get people in the habit of using implicit orderings).

Author: This will likely change as the MTK issues get fixed, hopefully, it will haven vaguely fast. Depending on those, this will change. Happy to keep this open until we know what syntax we can use.

Reviewer: Whatever syntax gets accepted there it will still make use of the symbolic for indexing. You can get the symbolic like
```julia
@unpack X = prob.f.sys
remake(prob; p = [X => b_values[i]])
```
I think.

Author: Tried this, but the returned problem is identical to the unmodified one. Think this was similar to one of the issues we had previously, I'll check if I can find what went wrong there.

Reviewer: My code had a typo and used X instead of b. Does it work with the right symbolic?

Author: Right, yeah, I should have caught that.

We use `@unpack` here now, but we still have to use indexing for the callback, as I don't think you can symbolically set index values for integrators...

end
nothing # hide
```
Here, `prob_func` takes three arguments:

>  - `prob`: The problem given to our `EnsembleProblem`, this is the problem that `prob_func` modified in each iteration.

Reviewer: ```suggestion
 - `prob`: The problem given to our `EnsembleProblem`, this is the problem that `prob_func` modifies in each iteration.
```

 - `i`: The number of this specific Monte Carlo iteration in the interval `1:trajectories`.
 - `repeat`: The repeat of this specific Monte Carlo simulation (We will ignore this argument in this brief overview).
In our case, for each Monte Carlo simulation, our `prob_func` takes our original `ODEProblem` and uses the `remake` function to change the parameter vector. Here, for the `i`th Monte Carlo simulation, the value of `b` is also the `i`th value of our `b_values` vector. Finally, we can simulate and plot our problem:
```@example ex1
eprob = EnsembleProblem(oprob; prob_func=prob_func)
esol = solve(eprob; trajectories=length(b_values))
plot(esol)
```

> Finally, it is worth noting that plot legends are disabled when plotting ensemble solutions. These can be re-enabled using the `legend` plotting keyword. However, when plotting a large number of trajectories, each will generate a label. Sometimes the best approach is to remove these and add a label manually:

Reviewer: ```suggestion
Note that plot legends are disabled when plotting ensemble solutions. These can be re-enabled using the `legend` plotting keyword. However, when plotting a large number of trajectories, each will generate a label. Sometimes the best approach is to remove these and add a label manually:
```

```
plot(esol; label="")
plot!([],[],label="X", legend=:best)
```

> ### Event handling using callbacks

Reviewer: Before we release Catalyst 13 this should be modified to use the symbolic event interface via `ReactionSystem`s (I'll add discrete event support).

Author: Yes, I kind of left this lose for now since there are currently several MTK updates underway that might change how we do this.

### Useful plotting options
Catalyst, just like DifferentialEquations, uses the Plots package for all plotting. For a detailed description of differential equation plotting, see [DifferentialEquations documentation on the subject](https://docs.sciml.ai/DiffEqDocs/stable/basics/plot/). Furthermore, the [Plots package documentation](https://docs.juliaplots.org/stable/) contains additional information and describes [a large number of plotting options](https://docs.juliaplots.org/stable/attributes/). Here follows a very short tutorial with a few useful options.

Let us consider the Brusselator model:
```@example ex3
brusselator = @reaction_network begin
    A, ∅ → X
    1, 2X + Y → 3X
    B, X → Y
    1, X → ∅
end
u0 = [:X => 1.0, :Y => 0.0]
tspan = (0.0,50.0)
p = [:A => 1.0, :B => 4.0]

oprob = ODEProblem(brusselator, u0, tspan, p)
sol = solve(oprob)
plot(sol)
```
If we want to plot only the `X` species, we can use the `idxs` command:
```@example ex3

> plot(sol; idxs=[brusselator.X], label="X(t)")

Reviewer: ```suggestion
@unpack X = brusselator
plot(sol; idxs=[X])
```
or give your system a name.

```
Here we use the `brusselator.X` notation to denote that we wish to plot the `X` species. The input to `idxs` is a vector listing all the species we wish to plot. If we wish to plot a single species, vector notation is not required and we could simply write `plot(sol; idxs=brusselator.X)`. 
(The plotting feature that automatically sets the label when using this interface is currently not working optimally, hence we manually set the label using the `label` option.)

Next, if we wish to plot a solution in phase space (instead of across time) we again use the `idxs` notation, but use `()` instead of `[]` when designating the species we wish to plot. Here, we plot the solution in `(X,Y)` space:
```@example ex3

> plot(sol; idxs=(brusselator.X,brusselator.Y), xguide="X", yguide="Y", label="(X,Y)")

Reviewer: ```suggestion
@unpack X,Y = brusselator
plot(sol; idxs=(X,Y))
```

Let's look at a single-component bistable self-activation model:
```@example ex1
using Catalyst, DifferentialEquations, Plots

rn = @reaction_network begin
    v0 + hill(X,v,K,n), ∅ --> X
    deg, X --> ∅
end
u0 = [:X => 0.0]
tspan = (0.0,1000.0)
p = [:v0 => 0.1, :v => 2.5, :K => 75.0, :n => 2.0, :deg => 0.01];

> sprob = SDEProblem(rn,u0,tspan,p)

Reviewer: ```suggestion
sprob = SDEProblem(rn, u0, tspan, p)
```

nothing # hide
```
we can then use our `SDEProblem` as input to an `EnsembleProblem`:
```@example ex1
eprob = EnsembleProblem(sprob)
```
The `EnsembleProblem` can now be used as input to the `solve` command. It has the same options as when simulating the `SDEProblem` directly, however, it has an additional argument `trajectories` to determine how many simulations should be performed. 
```@example ex1
esol = solve(eprob; trajectories=5)
```
This simulation is automatically multithreaded over all available threads. Please read [this documentation](https://docs.sciml.ai/DiffEqDocs/stable/features/ensemble/#EnsembleAlgorithms) for more information on parallelisation alternatives. The ensemble simulations can be plotted using the `plot` function, which by default displays all trajectories:
```@example ex1
plot(esol)
```

Sometimes when performing a large number of ensemble simulations, the plots get very dense. In these cases, the plot argument `linealpha` (which sets trajectory transparency) may be useful:
```@example ex1
esol = solve(eprob; trajectories=100)

> plot(esol)

Reviewer: This still isn’t using `linealpha`?

Author: Right

p = [:k => 1.0]
oprob = ODEProblem(rn, u0, tspan, p)

ps_cb_1 = PresetTimeCallback([3.0, 7.0], integ -> integ.u[1] += 5.0)
ps_cb_2 = PresetTimeCallback([5.0], integ -> integ.p[1] = 5.0)

sol = solve(deepcopy(oprob); callback=CallbackSet(ps_cb_1, ps_cb_2))
plot(sol)
```

The difference between the `PresetTimeCallback`s and the `DiscreteCallback`s and `ContiniousCallback`s is that the latter two allow the condition to be a function, permitting the user to give more general conditions for the callback to be triggered. An example could be a callback that triggers whenever a species surpasses some threshold value.

### Scaling the noise magnitude in the chemical Langevin equations
When using the CLE to generate SDEs from a CRN, it can sometimes be desirable to scale the magnitude of the noise terms. This can be done by introducing a *noise scaling parameter*. First, we simulate a simple two-state CRN model using the CLE:
```@example ex3
using Catalyst, StochasticDiffEq, Plots

rn_1 = @reaction_network begin
    (k1,k2), X1 <--> X2
end
u0 = [:X1 => 10.0, :X2 => 10.0]
tspan = (0.0, 10.0)
p_1 = [:k1 => 1.0, :k2 => 1.0]

sprob_1 = SDEProblem(rn_1, u0, tspan, p_1)
sol_1 = solve(sprob_1)

> plot(sol_1; idxs = 1, ylimit=(0.0, 20.0))

Reviewer: ```suggestion
plot(sol_1; idxs = 1, ylimit = (0.0, 20.0))
```

```
Here we can see that the `X` concentration fluctuations around a steady state of *X≈10.0*. 

Next, we wish to introduce a noise scaling parameter ,`η`. This will scale the noise magnitude so that for *η≈0.0* the system lacks noise (and its SDE simulations are identical to its ODE simulations) and for *η≈1.0* noise is not scaled (and SDE simulations are identical to as if no noise scaling was used). Setting *η<1.0* will reduce noise and *η>1.0* will increase noise. The syntax for setting a noise scaling parameter `η` is
```@example ex3
rn_2 = @reaction_network begin
    @parameters η
    (k1,k2), X1 <--> X2
end
u0 = [:X1 => 10.0, :X2 => 10.0]
tspan = (0.0, 10.0)
p_2 = [:k1 => 1.0, :k2 => 1.0, :η => 0.1]

sprob_2 = SDEProblem(rn_2, u0, tspan, p_2; noise_scaling = (@parameters η)[1])
```
Here, we first need to add `η` as a parameter to the system using the `@parameters η` option. Next, we pass the `noise_scaling = (@parameters η)[1]` argument to the `SDEProblem`. We can now simulate our system and confirm that noise is reduced:
```@example ex3
sol_2 = solve(sprob_2)

> plot(sol_2; idxs = 1, ylimit=(0.0, 20.0))

Reviewer: ```suggestion
plot(sol_2; idxs = 1, ylimit = (0.0, 20.0))
```

```

Finally, it is possible to set individual noise scaling parameters for each reaction of the system. Our model has two reactions (`X1 --> X2` and `X2 --> X1`) so we will use two noise scaling parameters, `η1` and `η2`. We use the following syntax:
```@example ex3
rn_3 = @reaction_network begin
    @parameters η1 η2
    (k1,k2), X1 <--> X2
end
u0 = [:X1 => 10.0, :X2 => 10.0]
tspan = (0.0, 10.0)
p_3 = [:k1 => 1.0, :k2 => 1.0, :η1 => 0.1, :η2 => 1.0]

sprob_3 = SDEProblem(rn_3, u0, tspan, p_3; noise_scaling = @parameters η1 η2)
```
plotting the results, we see that we have less fluctuation than for the first simulation, but more as compared to the second one (which is as expected):
```@example ex3
sol_3 = solve(sprob_3)

> plot(sol_3; idxs = 1, ylimit=(0.0, 20.0))

Reviewer: ```suggestion
plot(sol_3; idxs = 1, ylimit = (0.0, 20.0))
```

```

### Useful plotting options
Catalyst, just like DifferentialEquations, uses the Plots package for all plotting. For a detailed description of differential equation plotting, see [DifferentialEquations documentation on the subject](https://docs.sciml.ai/DiffEqDocs/stable/basics/plot/). Furthermore, the [Plots package documentation](https://docs.juliaplots.org/stable/) contains additional information and describes [a large number of plotting options](https://docs.juliaplots.org/stable/attributes/). Here follows a very short tutorial with a few useful options.

Let us consider the Brusselator model:
```@example ex4
using Catalyst, DifferentialEquations, Plots

brusselator = @reaction_network begin
    A, ∅ → X
    1, 2X + Y → 3X
    B, X → Y
    1, X → ∅
end
u0 = [:X => 1.0, :Y => 0.0]
tspan = (0.0, 50.0)
p = [:A => 1.0, :B => 4.0]

oprob = ODEProblem(brusselator, u0, tspan, p)
sol = solve(oprob)
plot(sol)
```
If we want to plot only the `X` species, we can use the `idxs` command:
```@example ex4
@unpack X = brusselator

> plot(sol; idxs=[X])

Reviewer: ```suggestion
plot(sol; idxs = [X])
```

```
Here we use `@unpack` to import `X` to the local scope, enabling us to use it for plotting. The input to `idxs` is a vector listing all the species we wish to plot. If we wish to plot a single species, vector notation is not required and we could simply write `plot(sol; idxs=X)`. 

Next, if we wish to plot a solution in phase space (instead of across time) we again use the `idxs` notation, but use `()` instead of `[]` when designating the species we wish to plot. Here, we plot the solution in `(X,Y)` space:
```@example ex4
@unpack X, Y = brusselator

> plot(sol; idxs=(X, Y))

Reviewer: ```suggestion
plot(sol; idxs = (X, Y))
```

```@example ex3
sol_3 = solve(sprob_3)
plot(sol_3; idxs=1, ylimit=(0.0,20.0))
```

### Useful plotting options
Catalyst, just like DifferentialEquations, uses the Plots package for all plotting. For a detailed description of differential equation plotting, see [DifferentialEquations documentation on the subject](https://docs.sciml.ai/DiffEqDocs/stable/basics/plot/). Furthermore, the [Plots package documentation](https://docs.juliaplots.org/stable/) contains additional information and describes [a large number of plotting options](https://docs.juliaplots.org/stable/attributes/). Here follows a very short tutorial with a few useful options.

Let us consider the Brusselator model:
```@example ex4
using Catalyst, DifferentialEquations, Plots

brusselator = @reaction_network begin
    A, ∅ → X
    1, 2X + Y → 3X
    B, X → Y
    1, X → ∅
end
u0 = [:X => 1.0, :Y => 0.0]
tspan = (0.0,50.0)
p = [:A => 1.0, :B => 4.0]

oprob = ODEProblem(brusselator, u0, tspan, p)
sol = solve(oprob)
plot(sol)
```
If we want to plot only the `X` species, we can use the `idxs` command:
```@example ex4
@unpack X = brusselator
plot(sol; idxs=[X])
```

> Here we use the `brusselator.X` notation to denote that we wish to plot the `X` species. The input to `idxs` is a vector listing all the species we wish to plot. If we wish to plot a single species, vector notation is not required and we could simply write `plot(sol; idxs=brusselator.X)`. 

Reviewer: This text needs to be updated since you unpack X now.

Author: Thanks for catching this one

> (The plotting feature that automatically sets the label when using this interface is currently not working optimally, hence we manually set the label using the `label` option.)

Reviewer: You can drop this comment since you can just use `X` now. Also I don't think it is broken, the problem is you didn't name your system. If you gave it a simple name the legend would look fine (i.e. should use name.X). Since you didn't give a name the name that is used is what `gensym` created.

Author: Right, fixed this as well.
